### PR TITLE
Set up APIService and expose first public methods

### PR DESCRIPTION
**Pull request by andrewatwood**
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature

- **What is the current behavior?** (You can also link to an open issue here)
No public methods exposed on the SDK client instance.

- **What is the new behavior (if this is a feature change)?**
Sets up the API service to make standardized API requests, and sets up the first public methods on the client via `AuthResource`, which contains a method `isSessionValid`

- **Testing**

1. Grab the `permSession` and `permMFA` cookies, and the API key from the request body on any `permanent.org/api`
2. Using the credentials you grabbed from the browser, `const perm = new Permanent()` and test that `perm.auth.isSessionValid()` resolves to `true`

### CODE REVIEW

**Comment by andrewatwood** [comment]
Also open to general structure and organization review from @slifty and @jasonaowen as well!

**Comment by xmunoz** [comment]
I don't see any new tests for this new public method. Can you add some?

**Comment by xmunoz** [comment]
I'd also like to see the functionality added here documented in the README before merging.

**Comment by andrewatwood** [comment]
Great point! Can definitely update the README.

**Comment by jasonaowen** [comment]
@andrewatwood some of the commits in this PR were made without a committer email address, which we should fix. The rest use your personal email address; is that what you intended? Make sure to [set your email address in git](https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address#setting-your-commit-email-address-in-git), and then I'd be happy to pair on rebasing to fix these commits!

File: src/lib/client.ts
Comment (jasonaowen) [inline src/lib/client.ts:8]: Are either `accountId` or `archiveId` used? They don't appear to be, but then I'm confused about how users are able to log in; is all the login data in `sessionToken`? Do we need both `sessionToken` and `accountId`?

File: src/lib/api/base.repo.spec.ts
Comment (jasonaowen) [inline src/lib/api/base.repo.spec.ts:65]: Might be clearer to use a matching descriptive name:

```suggestion
  t.context.csrfStore.setCsrf('oldcsrf');
```

(Me, in this review, eliminating all fun from the codebase :sweat_smile: )

Oogie Boogie

File: README.md
Comment (jasonaowen) [inline README.md:29]: Awesome docs!

**Review by jasonaowen — CHANGES_REQUESTED** [review]
Looking good!

I'll test this tomorrow, but first some feedback just from reading the code.

File: src/lib/client.ts
Comment (andrewatwood) [reply src/lib/client.ts:12]: It's a real API key to the extent that the Permanent system has real API keys, but it's not a _unique_ API key. It's the same client key that the web apps have used since day one, and should probably already have been rotated out anyway. But since there's no special permissions or access granted to this or any API key, and this key has already been in the wild via anyone peeking at outgoing requests their dev tools, it doesn't feel like a critical issue imo.

Now I do believe I meant to not actually commit it, for all the reasons and best practices you mentioned, but accidentally left it in 😅 or maybe I talked myself into leaving it since it's a useless key anyway? Who could know, it was eons ago!

File: src/lib/model/request-vo.ts
Comment (andrewatwood) [reply src/lib/model/request-vo.ts:7]: That's correct!

**Comment by andrewatwood** [comment]
@jasonaowen hmm was confused for a minute trying to figure that out, but realized those email-less were edits from the GitHub UI I made on 'main' to set up some Actions and other repo housekeeping stuff, and synced back in to this branch for the PR. Didn't realize it didn't have an email on the commits, seems bizarre for a default behavior...Still open to fixing that if you'd like!

File: src/lib/resources/auth.resource.ts
Comment (andrewatwood) [reply src/lib/resources/auth.resource.ts:25]: Oh duh! The optional chaining had me overthinking, because I was thinking about the potential `null` without realizing that I was already returning a `===` and not `SimpleVO.value`

File: src/lib/client.ts
Comment (andrewatwood) [reply src/lib/client.ts:8]: Yep, you're right, I was just starting to realize this! `sessionToken` covers all that server-side, and the `accountId` would only be needed in some use cases that this SDK definitely won't be covering at first (and most likely never will!)

File: src/lib/client.ts
Comment (slifty) [reply src/lib/client.ts:11]: Just to chime in, I hear you that there's nothing special about that key at this point, but there *could* be in future, right?  I mean, it won't happen, but in a year is it possible that someone might grant permission to that key?  It's a compromised key, but most of the time people assume keys are not compromised, so we're relying on institutional memory to mitigate that compromise (whereas rotating it out would just elminate the risk).

That said, agreed that it could be prioritized down the line.

Security aside, things like this should be loaded from a configurable variable rather than hard coding for all the reasons Jason mentioned!  +1 to that request.

Comment (andrewatwood) [reply src/lib/client.ts]: 100% onboard, making the change!  I initially didn't want to require an extra configuration param that's going to be identical for all users BUT as mentioned, even if security isn't the main issue, it is a good choice overall to move to a configurable option.

**Comment by jasonaowen** [comment]
Thanks for pairing with me on rebasing, @andrewatwood! I'll work on testing this now.

File: src/lib/client.ts
Comment (jasonaowen) [inline src/lib/client.ts:26]: While testing, I realized that since this is a boundary between our code and client code, we should probably validate our inputs here; we can't depend on the client code being in TypeScript or respecting our type definitions.

I believe this is purely a developer experience issue, and not a security issue; this code is running in their environment, and does not grant them the ability to do anything they couldn't already do by hitting our API directly.

Calling `new Permanent({})` should give some kind of useful error, like "`sessionToken` is a required parameter" or so. Currently it silently accepts the empty object, and then `isSessionValid()` returns `false` - which, while technically correct, does not make obvious the reason for that.

**Review by jasonaowen — APPROVED** [review]
Awesome work, @andrewatwood ! This looks good to me.

I tested it by building, then running in the REPL:

```sh
$ npm run build
$ nodejs
> // omitted: assign secrets to variables
> const permsdk = require('./build/main/');
> const perm = new permsdk.Permanent({sessionToken: permSession, mfaToken: permMFA, archiveId: 0, apiKey});
> perm.auth.isSessionValid().then((result) => console.log('Is session valid? ' + result));
Promise { <pending> }
Is session valid? true
```

There's a small UX issue below. Your call on if you want to add it to this PR, follow up in a separate PR, or not do it at all!

File: src/lib/client.ts
Comment (andrewatwood) [reply src/lib/client.ts:26]: Yes! I actually wanted to throw an error or something if params were missing but the fully strict ruleset wouldn't shut up about always being defined if I tried to check for missing params as they're all required in the interface, and then I forgot to investigate how best to do that with that restriction. Open to suggestions on that!

Comment (jasonaowen) [reply src/lib/client.ts:26]: > Yes! I actually wanted to throw an error or something if params were missing but the fully strict ruleset wouldn't shut up about always being defined if I tried to check for missing params as they're all required in the interface, and then I forgot to investigate how best to do that with that restriction. Open to suggestions on that!

@slifty found [Joi](https://joi.dev/), and we're using it in [upload-service](https://github.com/PermanentOrg/upload-service/) - see https://github.com/PermanentOrg/upload-service/pull/11 for our first use of it. It's powerful, and using it is really straightforward and readable - strong recommend!

Comment (jasonaowen) [reply src/lib/client.ts:26]: Captured in https://github.com/PermanentOrg/node-sdk/issues/7.

**Pull request merged**
